File: src/main/java/com/buffer/lorena/bot/events/MessageHandler.java
package com.buffer.lorena.bot.events;

import com.buffer.lorena.bot.service.LorenaService;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.message.MessageCreateEvent;
import org.javacord.api.listener.message.MessageCreateListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Locale;

/**
 * The type Message handler.
 */
@Component
public class MessageHandler implements MessageCreateListener {
    private final Logger logger = LogManager.getLogger(MessageHandler.class);
    private static final String LORENA_TEXT = "lorena";

    private LorenaService lorenaService;

    /**
     * Instantiates a new Message handler.
     */
    public MessageHandler() {
    }

    /**
     * Instantiates a new Message handler.
     *
     * @param lorenaService     the lorena service
     */
    @Autowired
    public MessageHandler(LorenaService lorenaService) {
        this.lorenaService = lorenaService;
    }

    /**
     * On message create.
     *
     * @param event the event
     */
    @Override
    public void onMessageCreate(MessageCreateEvent event) {
        logger.info("message received in server {}: {}", event.getServer().map(Server::getName).get(), event.getMessageContent());
        String[] parsedMessage = event.getMessageContent().split(" ");
        if (parsedMessage[0].equalsIgnoreCase("!lore") || parsedMessage[0].equalsIgnoreCase("!lorebot")) {
            switch (parsedMessage[1].toLowerCase(Locale.ROOT)){
                case "ping":
                    event.getChannel().sendMessage("Pong!");
                    break;
                case "votes":
                    if(event.getMessageAuthor().isServerAdmin()) {
                        try {
                            int uvt = Integer.parseInt(parsedMessage[2]);
                            event.getServer().ifPresent(server -> {
                                this.lorenaService.changeServerUserVoteThreshold(server, uvt);
                                event.addReactionsToMessage("✅");
                            });
                        } catch (Exception e) {
                            event.addReactionsToMessage("❌");
                        }
                    }
                    break;
                case "setlorechannel":
                    if(event.getMessageAuthor().isServerAdmin()) {
                        try {
                            Long newLoreChannelId = Long.parseLong(parsedMessage[2].substring(2, parsedMessage[2].length()-1));
                            this.lorenaService.setServerLoreChannel(event.getServer().get(), newLoreChannelId);
                            event.addReactionsToMessage("✅");
                        } catch (Exception e) {
                            event.addReactionsToMessage("❌");
                        }
                    }
                    break;
                case "dolore":
                    this.lorenaService.sendRandomLore(event);
                    break;
                default:
                    event.getChannel().sendMessage("fuck off");
            }
        } else if(event.getMessageContent().toLowerCase(Locale.ROOT).contains(LORENA_TEXT)){
            this.lorenaService.sendRandomLore(event);
        }
    }
}

File: src/main/kotlin/com/buffer/lorena/events/MessageHandler.kt
package com.buffer.lorena.events

import org.javacord.api.listener.message.MessageCreateListener
import com.buffer.lorena.bot.service.LorenaService
import com.buffer.lorena.service.RedditService
import com.buffer.lorena.service.UnitConversionService
import org.apache.logging.log4j.LogManager
import org.javacord.api.entity.server.Server
import org.javacord.api.event.message.MessageCreateEvent
import org.springframework.stereotype.Component
import java.util.Locale

/**
 * The type Message handler.
 */
@Component
class MessageHandler(
        private val lorenaService: LorenaService,
        private val redditService: RedditService,
        private val unitConversionService: UnitConversionService,
) : MessageCreateListener {
    private val logger = LogManager.getLogger(MessageHandler::class.java)

    /**
     * On message create.
     *
     * @param event the event
     */
    override fun onMessageCreate(event: MessageCreateEvent) {
        logger.info("message received in server {}: {}", event.server.map { obj: Server -> obj.name }.get(), event.messageContent)

        val parsedMessage = event.messageContent.split(" ")

        when {
            (parsedMessage[0].equals("!lore", ignoreCase = true)
                    || parsedMessage[0].equals("!lorebot", ignoreCase = true))
                    && !event.messageAuthor.isBotUser ->
                when (parsedMessage[1].lowercase(Locale.ROOT)) {
                    "ping" -> event.channel.sendMessage("Pong!")
                    "votes" -> lorenaService.changeUserVoteThreshold(event, parsedMessage[2])
                    "setlorechannel" -> lorenaService.setServerLoreChannel(event, parsedMessage[2])
                    "dolore" -> lorenaService.sendRandomLore(event)
                    else -> event.channel.sendMessage("fuck off")
                }
            // TODO How come?
//            parsedMessage.any { it.equals(LORENA_TEXT, true) } -> lorenaService.sendRandomLore(event)
            event.messageContent.contains(LORENA_TEXT) -> lorenaService.sendRandomLore(event)
        }

        parsedMessage.filter { it.isReddit() }.forEach {
            redditService.sendLink(it, event)
        }

        unitConversionService.parseMessage(event.messageContent, event)
    }

    private fun String.isReddit(): Boolean = this.matches(REDDIT_REGEX)

    companion object {
        private const val LORENA_TEXT = "lorena"
        private val REDDIT_REGEX = """\b([ru])/([a-zA-Z0-9-_]{3,21})\b""".toRegex()
    }
}